lib/diskoverflow/map.go

```diff
@@ -0,0 +1,315 @@
+	values map[string]Value
+	bytes  int
+}
+
+func (o *memoryMap) add(k string, v Value) {
+	o.values[k] = v
+	o.bytes += v.Bytes()
+}
+
+func (o *memoryMap) Bytes() int {
+	return o.bytes
+}
+
+func (o *memoryMap) Close() {
+	o.values = nil
+}
+
+func (o *memoryMap) Get(k string, v Value) bool {
+	nv, ok := o.values[k]
+	if !ok {
+		return false
+	}
+	v.Copy(nv)
+	return true
+}
+
+func (o *memoryMap) Items() int {
+	return len(o.values)
+}
+
+func (o *memoryMap) Pop(k string, v Value) bool {
+	ok := o.Get(k, v)
+	if !ok {
+		return false
+	}
+	delete(o.values, k)
+	o.bytes -= v.Bytes()
+	return true
+}
+
+func (o *memoryMap) Delete(k string) {
+	v, ok := o.values[k]
+	if !ok {
+		return
+	}
+	delete(o.values, k)
+	o.bytes -= v.Bytes()
+}
+
+type iteratorValue struct {
+	k string
+	v Value
+}
+
+type memMapIterator struct {
+	values  map[string]Value
+	ch      chan iteratorValue
+	stop    chan struct{}
+	current iteratorValue
+	parent  iteratorParent
+}
+
+func (o *memoryMap) newIterator(p iteratorParent) MapIterator {
+	i := &memMapIterator{
+		ch:     make(chan iteratorValue),
+		stop:   make(chan struct{}),
+		parent: p,
+		values: o.values,
+	}
+	go i.iterate()
+	return i
+}
+
+func (i *memMapIterator) iterate() {
+	for k, v := range i.values {
+		select {
+		case <-i.stop:
+			break
+		case i.ch <- iteratorValue{k, v}:
+		}
+	}
+	close(i.ch)
+}
+
+func (i *memMapIterator) Key() string {
+	return i.current.k
+}
+
+func (i *memMapIterator) Value(v Value) {
+	v.Copy(i.current.v)
+}
+
+func (i *memMapIterator) Next() bool {
+	var ok bool
+	i.current, ok = <-i.ch
+	return ok
+}
+
+func (i *memMapIterator) Release() {
+	close(i.stop)
+	i.parent.released()
+}
+
+type diskMap struct {
+	db    *leveldb.DB
+	bytes int
+	dir   string
+	len   int
+}
+
+func newDiskMap(location string) *diskMap {
+	// Use a temporary database directory.
+	tmp, err := ioutil.TempDir(location, "overflow-")
+	if err != nil {
+		panic("creating temporary directory: " + err.Error())
+	}
+	db, err := leveldb.OpenFile(tmp, &opt.Options{
+		OpenFilesCacheCapacity: 10,
+		WriteBuffer:            512 << 10,
+	})
+	if err != nil {
+		panic("creating temporary database: " + err.Error())
+	}
+	return &diskMap{
+		db:  db,
+		dir: tmp,
+	}
+}
+
+func (o *diskMap) add(k string, v Value) {
+	o.addBytes([]byte(k), v)
+	o.bytes += v.Bytes()
+}
+
+func (o *diskMap) addBytes(k []byte, v Value) {
+	if err := o.db.Put(k, v.Marshal(), nil); err != nil {
+		panic("writing to temporary database: " + err.Error())
+	}
+	o.len++
+}
+
+func (o *diskMap) Close() {
+	o.db.Close()
+	os.RemoveAll(o.dir)
+}
+
+func (o *diskMap) Bytes() int {
+	return o.bytes
+}
+
+func (o *diskMap) Get(k string, v Value) bool {
+	d, err := o.db.Get([]byte(k), nil)
+	if err != nil {
+		return false
+	}
+	v.Unmarshal(d)
+	return true
+}
+
+func (o *diskMap) Items() int {
+	return o.len
+}
+
+func (o *diskMap) Pop(k string, v Value) bool {
+	ok := o.Get(k, v)
+	if ok {
+		_ = o.db.Delete([]byte(k), nil)
```

Comment:
So, the error handling here and in Delete()... We've verified that the item exists, in the Get(), so Delete() should not fail. If it does, it's an I/O error or something. Do we really want to swallow that? Here maybe, just to keep chugging, but in the Delete() case below I would not. It should be sufficient there to do the delete and if the error is non-nil and not DoesNotExist (or whatever that is exactly), panic etc?

Reply:
Right, and I think we even need to panic here as well, otherwise there's unexpected behaviour if the caller depends on a successive `Pop` for the same key to return false.